File: src/javascript/algorithms/strings/countBinarySubstrings/countBinarySubstrings.js
/**
 * Count Binary Substrings (LeetCode 696)
 *
 * Return the number of substrings with equal consecutive 0s and 1s.
 *
 * Approach: Group Counting
 * TC: O(n)  SC: O(1)
 */

function countBinarySubstrings(s) {
    let prev = 0, curr = 1, count = 0;

    for (let i = 1; i < s.length; i++) {
        if (s[i] === s[i - 1]) {
            curr++;
        } else {
            count += Math.min(prev, curr);
            prev = curr;
            curr = 1;
        }
    }

    count += Math.min(prev, curr);
    return count;
}

const testCases = [
    { s: "00110011", expected: 6, label: "groups [2,2,2,2]" },
    { s: "10101",    expected: 4, label: "groups [1,1,1,1,1]" },
    { s: "00011",    expected: 2, label: "groups [3,2]" },
    { s: "0",        expected: 0, label: "single char" },
    { s: "01",       expected: 1, label: "groups [1,1]" },
    { s: "000111",   expected: 3, label: "groups [3,3]" },
];

testCases.forEach(({ s, expected, label }) => {
    const result = countBinarySubstrings(s);
    const status = result === expected ? "PASS" : "FAIL";
    console.log(`  [${status}] "${s}" (${label}): ${result} (expected ${expected})`);
});

File: src/python/algorithms/strings/countBinarySubstrings/count_binary_substrings.py
"""
Count Binary Substrings (LeetCode 696)

Return the number of substrings with equal consecutive 0s and 1s.

Approach: Group Counting
TC: O(n)  SC: O(1)
"""


def count_binary_substrings(s: str) -> int:
    prev = 0
    curr = 1
    count = 0

    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            curr += 1
        else:
            count += min(prev, curr)
            prev = curr
            curr = 1

    count += min(prev, curr)
    return count


if __name__ == "__main__":
    test_cases = [
        ("00110011", 6, 'groups [2,2,2,2]'),
        ("10101",    4, 'groups [1,1,1,1,1]'),
        ("00011",    2, 'groups [3,2]'),
        ("0",        0, 'single char'),
        ("01",       1, 'groups [1,1]'),
        ("000111",   3, 'groups [3,3]'),
    ]

    for s, expected, label in test_cases:
        result = count_binary_substrings(s)
        status = "PASS" if result == expected else "FAIL"
        print(f"  [{status}] \"{s}\" ({label}): {result} (expected {expected})")
